ENH: Annotation Storage Additions For Visualization

measty opened the pull request:
This PR adds some enhancements to storage.py that are useful when using an AnnotationStore for visualization purposes.

Main changes:

Support for adding annotations from a hovernet-style .dat through a utility function

Addition of 'area' field to table in SQliteStore. Annotations are returned sorted by area, and supports querying for annotations satisfying an area threshold, useful for zoomed-out rendering where annotations which are too small to display can be excluded.
After the discussions in tiatoolbox meeting, functions are added to allow to add/remove the area column to save space if its not expected to be needed.
Sorting annotations efficiently by area (can create an index on it) allows to effficiently deal with potentially overlapping annotations by drawing geometries of smallest area last. Eg

Unsorted rendering:
![before_area_sort](https://user-images.githubusercontent.com/20169086/185522132-9e902b7b-83a3-4860-84b6-332901a96b70.png)

Sorted rendering:
![after_area_sort](https://user-images.githubusercontent.com/20169086/185522211-6d227397-8fca-4ce0-9efb-cb7c2b2aa66a.png)

~~Addition of cacheable versions of query, bquery. The intention with these is to potentially be more efficient when viewing the same area of screen and modifying various options controlling how annotations are rendered (perhaps changing color mapping, or deselecting objects of specific types to be renderered). As in these cases we are displaying the same annotations just in a different way, we dont need to re-load all the annotations for each tile from the store; we can just load the cached annotations and then filter/render them differently.
Still needs a bit of testing to see how much this actually helps/if its worth using~~ (caching aspects of the PR removed for now after feedback, may revisit this later in a separate PR)

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1321,19 +1557,22 @@ def clear(self) -> None:
+                Path object.
+            con (sqlite3.Connection):
+                The sqlite3 database connection.
+    <<<<<<< HEAD
```
Delete 

<<<<<<< HEAD
    =======
    >>>>>>> b4cf2ebc5ad0c367d425c50b631748de42434e4f

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1303,6 +1522,23 @@ def to_dataframe(self) -> pd.DataFrame:
         return pd.json_normalize(features).set_index("key")
 
+    def translate_db(self, x: float, y: float):
+        """Translate all annotations by the given vector. Useful for
```
```suggestion
        """Translate all annotations by the given vector. 
        
        Useful for
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1895,20 +2154,226 @@ def query(
+        compress_type=None,
+        min_area=None,
+    ) -> sqlite3.Cursor:
+        """Common query construction logic for `cached_query` and `cached_bquery`. Similar
```
```suggestion
        """Common query construction logic for `cached_query` and `cached_bquery`. 
        
        Similar
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1895,20 +2154,226 @@ def query(
+        where: Callable[[Dict[str, Any]], bool] = None,
+        min_area=None,
+    ) -> Dict[str, Annotation]:
+        """Query the store for annotations. Similar to `bquery` but
```
```suggestion
        """Query the store for annotations. 
        
        Similar to `bquery` but
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1895,20 +2154,226 @@ def query(
+        where: Callable[[Dict[str, Any]], bool] = None,
+        min_area=None,
+    ) -> List[Annotation]:
+        """Query the store for annotations. Similar to `query` but
```
```suggestion
        """Query the store for annotations. 
        
        Similar to `query` but
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1136,19 +1142,232 @@ def _load_cases(
+        Returns:
+            AnnotationStore:
+                A new annotation store with the annotations loaded from the file.
+        """
```
```suggestion

        """
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1136,19 +1142,232 @@ def _load_cases(
+        Returns:
+            AnnotationStore:
+                A new annotation store with the annotations loaded from the file.
+        """
```
```suggestion

        """
```

John-P commented:
I will have to provide longer feedback in future but just making a few quick notes before our meeting. I have a few concerns:
1. I don't think it wise to add more columns to the database. When doing the extensive analysis and use case argument for using this SQLite store, storage space was a big concern, especially with large sets of annotations. I would suggest using a temporary table to keep areas for visualization purposes. This will compute them once when the database is opened and not write them to disk, taking up more space. Alternatively, you could use the bounding box area which can be obtained in a select statement from the rtree index already.
2. Beware caching. It can be full of pitfalls from memory leaks to cache invalidation issues. The built in functools LRU cache has been known to have issues with memory leaks. I would suggest moving this up the stack out of the storage class itself and into the visualization layer also. There is the issue of if someone update the database from another process. How does this get propagated etc.

measty commented:
> 



> I will have to provide longer feedback in future, but just making a few quick notes before our meeting. I have a few concerns:
> 
> 1. I don't think it wise to add more columns to the database. When doing the extensive analysis and use case argument for using this SQLite store, storage space was a big concern, especially with large sets of annotations. I would suggest using a temporary table to keep areas (and other region props) for visualization purposes. This will compute them once when the database is opened and not write them to disk, taking up more space. Alternatively, you could use the bounding box area, which can be obtained in a select statement from the rtree index already without the need to extra columns. Additionally, I think that area might be calculated already when the WKB is parsed. Have you checked this? I.e. does it offer any speedup to have this saved somehwere vs obtaining it from the shapely object?
> 2. Beware caching. It can be full of pitfalls from memory leaks to cache invalidation issues. The built-in functools LRU cache has been known to have issues with memory leaks. I would suggest moving this up the stack out of the storage class itself and into the visualization layer also. There is the issue of if someone updates the database from another process. How does this get propagated etc.
> 3. Dat is a very generic format. Calling it `from_dat` is confusing as it is specific to our toolbox / hovernet rather than being able to use any dat file. Perhaps `from_hovernet_dat` would be better. Also, this adds the first tight coupling between the annotation store and another part of the toolbox. I think we should avoid this. Especially, if annotation store is to be broken out into a separate package in the future. I can't think of an elegant solution right now but perhaps some loader function which takes a store and a dat file to load into the database would be better than baking it into the storage class to avoid this coupling for now.

1. I have changed the area column to an interger to save space, as that is sufficient for what is needed. The impact on filesize was around 3-4%. My concerns with the alternate solutions are that BBox area is not sufficient, as that gives a large area to line annotations, which really need their correct zero area as they will show up well plotted over a region annotation, but not vice-versa. And the temporary table solution, I am concerned that calculating the area of all annotations and creating an index on it on load will lengthen the already noticable delay on opening a new store. My first attempt at using area originally had all the filtering/sorting by area outside of the storage class, I moved it into the db as its more efficient to have it there so that you can have an index on it, and so area constraints can be incorporated directly in the query so that you dont even return from the database the annotations you dont plan to render.
2. I've removed the cached querys for now, can potentially rethink how its done later. 
3. I've moved the code to create a store from hovernet style .dat into a standalone function in tiatoolbox\utils\misc.py

measty commented:
I've added functions to add/drop the area column as discussed. Also finished adding tests, so marked it as ready for review.

John-P commented:
@measty It has since occurred to me that we may need to add a column for annotation type (e.g. an integer which corresponds to an enum) for filtering annotation etc. This could also be used instead of / as a fallback for area. I.e.g points/lines get 0 (no computation needed) and polygons get bounding box from the rtree at query. This should be very fast as it is a simple multiplication performed by the query engine. It may in fact be more performant than using the actual area on cold startup and have negligible difference for warm queries also. Can we benchmark this?

shaneahmed commented on tests/test_annotation_stores.py:
```diff
@@ -755,6 +838,36 @@ def test_from_geojson_path(fill_store, tmp_path, store_cls):
+        """Test translating a store."""
+
+        def test_translation(geom):
+            """Performas a translation of input geometry."""
```
```suggestion
            """Performs a translation of input geometry."""
```

shaneahmed commented on tests/test_utils.py:
```diff
@@ -1366,3 +1369,104 @@ def test_detect_gpu():
+
+
+def test_from_dat_typedict(tmp_path):
+    """Test generating an annotation store from a .dat file with a typedict."""
```
```suggestion
    """Test generating an annotation store from a .dat file with a type dict."""
```

shaneahmed commented on tests/test_utils.py:
```diff
@@ -1366,3 +1369,104 @@ def test_detect_gpu():
+    assert np.rint(poly.geometry.centroid.y) == 150
+
+
+def test_from_multihead_dat(tmp_path):
```
```suggestion
def test_from_multihead_dat(tmp_path):
```

shaneahmed commented on tests/test_utils.py:
```diff
@@ -1366,3 +1369,104 @@ def test_detect_gpu():
+        "A": head_a,
+        "B": head_b,
+        "resolution": 0.5,
+        "othermetadata": {"foo": "bar"},
```
```suggestion
        "other_meta_data": {"foo": "bar"},
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1117,18 +1118,84 @@ def _load_cases(
+    def from_geojson(
+        cls,
+        fp: Union[IO, str],
+        scale_factor=1,
```
```suggestion
        scale_factor: int=1,
```

shaneahmed replied:
please add typehints

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1117,18 +1118,84 @@ def _load_cases(
+        cls,
+        fp: Union[IO, str],
+        scale_factor=1,
+        relative_to=None,
```
```suggestion
        relative_to:Tuple[float, float]=None,
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1117,18 +1118,84 @@ def _load_cases(
+        self,
+        fp: Union[IO, str],
+        scale_factor=1,
+        relative_to=None,
```
Add typehints

shaneahmed replied:
Please be consistent with typehints.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1117,18 +1118,84 @@ def _load_cases(
-            store.append(Annotation(geometry, properties))
-        return store
+
+        anns = [
```
```suggestion
annotations = [
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1318,7 +1401,6 @@ class SQLiteMetadata(MutableMapping):
             Path object.
         con (sqlite3.Connection):
             The sqlite3 database connection.
```
```suggestion
new line
"""
```

shaneahmed commented on tiatoolbox/utils/misc.py:
```diff
@@ -849,3 +854,157 @@ def select_cv2_interpolation(scale_factor):
+
+
+def store_from_dat(
+    fp: Union[IO, str],
```
Add type hints

shaneahmed commented on tiatoolbox/utils/misc.py:
```diff
@@ -849,3 +854,157 @@ def select_cv2_interpolation(scale_factor):
+    cls: AnnotationStore = SQLiteStore,
+) -> "AnnotationStore":
+    """Load annotations from a hovernet-style .dat file.
+    Args:
```
new line

shaneahmed commented on tiatoolbox/utils/misc.py:
```diff
@@ -849,3 +854,157 @@ def select_cv2_interpolation(scale_factor):
+    """Helper function to create list of Annotation objects from a
+    hovernet-style dict of segmentations, mapping types using typedict
+    if provided.
+    """
```
new line

shaneahmed commented on tiatoolbox/utils/misc.py:
```diff
@@ -849,3 +854,157 @@ def select_cv2_interpolation(scale_factor):
+
+
+def anns_from_hoverdict(data, props, typedict, relative_to, scale_factor):
+    """Helper function to create list of Annotation objects from a
```
first line should be sentence
add help for args.

shaneahmed commented on tiatoolbox/utils/misc.py:
```diff
@@ -849,3 +854,157 @@ def select_cv2_interpolation(scale_factor):
+def make_default_dict(data, subcat):
+    """Helper function to create a default typedict if none is provided.
+
+    The unique types in the data are given a prefix to differentiate
```
Args:?

shaneahmed commented on tiatoolbox/utils/misc.py:
```diff
@@ -849,3 +854,157 @@ def select_cv2_interpolation(scale_factor):
+                'head2': {1: 'Gland', 2: 'Lumen', 3: ...}, ...}.
+        relative_to [float, float]:
+            The x and y coordinates to use as the origin for the annotations.
+    """
```
new line
returns:?

measty replied:
That function doesnt return anything

shaneahmed replied:
ok. please can you add new line?

measty commented:
> Thanks @measty. Please can you address the comments?
@shaneahmed the docstring-related comments have been addressed

measty commented:
> @measty It has since occurred to me that we may need to add a column for annotation type (e.g. an integer which corresponds to an enum) for filtering annotations etc. This could also be used instead of / as a fallback for area. E.g points/lines get 0 (no computation needed) and polygons get bounding box from the rtree at query. This should be very fast as it is a simple multiplication performed by the query engine. It may in fact be more performant than using the actual area on cold startup (because calculating actual area is slow vs fast bounding box) and may have negligible difference for warm queries also. Can we benchmark this?
> 
> If you still want the actual area, the bounding box could be used if the area column is missing at start up (or first query) to reduce lag time. Meanwhile, in the background you could perform a query to add the missing column to avoid start up lag. However, we may find that this combo of annotation type + bounding box is enough.

@John-P I tried adding a custom function:

```
def get_area_approx(objtype, min_x, min_y, max_x, max_y):
       """Function to get the approx. area of a geometry."""
       if objtype in ("Point", "LineString"):
           return 0
       return (max_x - min_x) * (max_y - min_y)
```

And using that for the area based sorting/filtering. I did a small bit of benchmarking, rendering all the possible tiles in the generator, and just the first 50 to simulate its effect after initial load. 
All tiles: 279s using above function, 275s using area column
First 50: 4.7s using above fn, 3.06 using area column.

So, it seems to be a reasonable fallback, and doesn't degrade performance much apart from a bit of a hit to the initial lag.

shaneahmed commented on tests/test_annotation_stores.py:
```diff
@@ -755,6 +849,36 @@ def test_from_geojson_path(fill_store, tmp_path, store_cls):
 
+    @staticmethod
+    def test_from_geojson_path_transform(fill_store, tmp_path, store_cls):
+        """Test loading from geojson with a transform"""
```
```suggestion
        """Test loading from geojson with a transform."""
```

shaneahmed commented on tests/test_annotation_stores.py:
```diff
@@ -755,6 +849,36 @@ def test_from_geojson_path(fill_store, tmp_path, store_cls):
+    @staticmethod
+    def test_transform_db(fill_store, tmp_path, store_cls):
+        """Test translating a store."""
+
```
```suggestion
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1117,18 +1118,84 @@ def _load_cases(
+    ) -> "AnnotationStore":
+        """Create a new database with annotations loaded from a geoJSON file.
+        Args:
+            fp (Union[IO, str, Path]):
```
Please can you check if we are using file_path variable in misc? It would be better to be consistent. Also, fp is not clear.

measty replied:
I have used fp in everything I have added, to be consistent with what was already used throughout storage.py

measty replied:
Have replaced fp with full_path

John-P replied:
It's supposed to be a common shorthand for file path here (rather than full path). Happy with file_path instead.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1117,18 +1118,84 @@ def _load_cases(
+
+    def add_from_geojson(
+        self,
+        fp: Union[IO, str],
```
```suggestion
        file_path: Union[IO, str],
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1117,18 +1118,84 @@ def _load_cases(
+                at non-baseline resolution.
+            relative_to [float, float]:
+                The x and y coordinates to use as the origin for the annotations.
+        """
```
```suggestion

        """
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1117,18 +1118,84 @@ def _load_cases(
+            relative_to [float, float]:
+                The x and y coordinates to use as the origin for the annotations.
+        """
+
```
```suggestion
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1288,6 +1355,22 @@ def to_dataframe(self) -> pd.DataFrame:
+            transform (callable[Geometry, Geometry]):
+                A function that takes a geometry and returns a new
+                transformed geometry.
+        """
```
```suggestion

        """
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1318,7 +1401,6 @@ class SQLiteMetadata(MutableMapping):
         con (sqlite3.Connection):
             The sqlite3 database connection.
-
     """
```
```suggestion

    """
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1382,6 +1464,7 @@ def __init__(
         connection: Union[Path, str, IO] = ":memory:",
         compression="zlib",
```
Please be consistent with typehints.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1500,11 +1600,14 @@ def register_custom_function(
+                properties TEXT,         -- JSON properties
+                area INTEGER NOT NULL    -- Area (for ordering)
             )
             """
```
```suggestion

            """
```

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1677,7 +1785,7 @@ def _append(self, key: str, annotation: Annotation, cur: sqlite3.Cursor) -> None
-                    :cx, :cy, :geometry, :properties
+                    :cx, :cy, :geometry, :properties, :area
                 )
                 """,
```
```suggestion

                """,
```

shaneahmed commented on tiatoolbox/utils/misc.py:
```diff
@@ -849,3 +854,196 @@ def select_cv2_interpolation(scale_factor):
+    typedict=None,
+    relative_to=None,
+) -> None:
+    """Add annotations from a .dat file to an existing store. Make
```
```suggestion
    """Add annotations from a .dat file to an existing store. 
    
    Make a best effort to create valid shapely geometries from provided contours.
```

shaneahmed commented on tiatoolbox/utils/misc.py:
```diff
@@ -849,3 +854,196 @@ def select_cv2_interpolation(scale_factor):
+    relative_to=None,
+) -> None:
+    """Add annotations from a .dat file to an existing store. Make
+    a best effort to create valid shapely geometries from provided contours.
```
```suggestion
```

shaneahmed reviewed (approved):
Looks good to me @measty.

Please make sure that your address the comments. My main comments are on type hints and consistency of variable names. I will wait for @John-P's review before merge.

John-P commented:
One thing to note is that if any changes are being made to the table schema, this must be done carefully to avoid breaking things too much. E.g. we need:
- To document changes in the docstrings.
- To increment the version number.
- To add checks for version number.
- Handle old versions / upgrade if required.

John-P commented:
Just copying over comments here from out last team chat, If we change the type column to be an int/enum instead of text then this would be a breaking change. However, it could be a significant space-saving and hopefully more than account for adding a true area column if this is required. By using the type, we can also get a pseudo area, i.e. 0 for lines and points and the bounding-box area for polygons. Additionally, we could probably remove the cx and cy columns. They are currently used to store a 'representative point', i.e. a point near the center of mass but within the shape. We are currently not using this. Additionally, a bounding box center point can be obtained from the rtree index.

John-P commented on tests/test_annotation_stores.py:
```diff
@@ -15,7 +15,7 @@
 import pytest
 from shapely import affinity
-from shapely.geometry import Polygon
+from shapely.geometry import MultiPoint, Polygon
```
Can I just check why you are using MultiPoint here? I personally can't see many reasons to use this over multiple Point annotations as it makes spatial queries not work as well. Is this just to test that they do work?

John-P commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1071,80 +1072,147 @@ def to_geodict(self) -> Dict[str, Any]:
+                The scale factor to use when loading the annotations. All coordinates
+                will be multiplied by this factor to allow import of annotations saved
+                at non-baseline resolution.
+            relative_to [float, float]:
```
```suggestion
            origin [float, float] = (0, 0):
```

I think 'origin' is slightly less ambiguous / more clear than 'relative_to' IMO.

measty replied:
Origin would be clearer, i'll change it to that

John-P commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1071,80 +1072,147 @@ def to_geodict(self) -> Dict[str, Any]:
+    def from_geojson(
+        cls,
+        full_path: Union[IO, str],
+        scale_factor: float = 1,
```
Maybe overkill but perhaps list of transforms could be given to allow other things just as inverting the origin / mirroring or rotating etc. Similar to the coordinateTransformations in NGFF v0.4.

John-P commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1071,80 +1072,147 @@ def to_geodict(self) -> Dict[str, Any]:
+
+        """
+
+        def transform_geom(geom):
```
```suggestion
        def transform_geomety(geom):
```

'geom' is not used anywhere else. I think just using the actual word is more clear.

John-P commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1071,80 +1072,147 @@ def to_geodict(self) -> Dict[str, Any]:
+        def transform_geom(geom):
+            """Helper function to transform a geometry if needed."""
+            if relative_to is not None:
+                # transform coords to be relative to given pt.
```
```suggestion
                # Transform coords to be relative to given point.
```

measty replied:
its just a change of origin, but worded badly. I'll clarify

John-P commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1071,80 +1072,147 @@ def to_geodict(self) -> Dict[str, Any]:
 
         Args:
-             fp (IO):
+             full_path (IO):
```
This name doesn't make sense. It is not for a path but a file-like IO object. 
`fp` was chosen to be consistent with the Python standard library e.g. json: https://github.com/python/cpython/blob/c7662420d62e04eefd99d7cdbcc74f72e68f3a6a/Lib/json/__init__.py#L120

Although admittedly this also works with a file path. and should be a union of `IO` and `pathlib.Path`.

Full path also doesn't make sense as it works with non-full paths i.e. relative paths.

John-P replied:
```suggestion
             full_path (IO|pathlib.Path|str):
```

John-P commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1170,12 +1238,12 @@ def write_geojson_to_file_handle(file_handle: IO):
 
         return self._dump_cases(
-            fp=fp,
+            full_path=full_path,
```
Again full path doesn't make sense as a name here. It can be an IO object or a file path (full or relative etc.).

measty replied:
i've changed this back to fp again

John-P commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1288,6 +1356,23 @@ def to_dataframe(self) -> pd.DataFrame:
         )
         return pd.json_normalize(features).set_index("key")
 
+    def transform_db(self, transform: Callable[[Geometry], Geometry]) -> None:
```
If this is transforming the store don't put db in the name. Is this being implemented for all stores to keep the API consistent?

measty replied:
It is implemented in the AnnotionStore base class so will be available in all stores

John-P replied:
Great, thanks for clarifying. I missed that it was in the base class.

John-P commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1288,6 +1356,23 @@ def to_dataframe(self) -> pd.DataFrame:
         )
         return pd.json_normalize(features).set_index("key")
 
+    def transform_db(self, transform: Callable[[Geometry], Geometry]) -> None:
```
```suggestion
    def transform(self, transform: Callable[[Geometry], Geometry]) -> None:
```

John-P reviewed (changes requested):
I've make a few comments / suggestions of things. Mostly just suggesting alternative naming for things.

John-P commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1071,80 +1072,147 @@ def to_geodict(self) -> Dict[str, Any]:
+        print(f"added {len(annotations)} annotations")
+        self.append_many(annotations)
+
+    def to_geojson(self, full_path: Optional[IO] = None) -> Optional[str]:
```
```suggestion
    def to_geojson(self, full_path: Optional[IO|pathlib.Path|str] = None) -> Optional[str]:
```

`|` can be used instead of `Union[IO, str, Path]`, not sure if this is in python 3.7 though so might have to be `Union[IO, str, Path]`

measty replied:
yes, older versions of python weren't happy with | so will have to be union for now

John-P commented on tiatoolbox/annotation/storage.py:
```diff
@@ -1371,17 +1456,24 @@ class SQLiteStore(AnnotationStore):
+    1.0.0:
+        Initial version.
+    1.0.1 (07/10/2022):
+        Added support for Area column and queries sorted/filtered by area.
```
Just wanted to clarify how this is handling a store when loaded which does not have an area column> Is it just ignoring it and raising an error if you try to order on area? Or is it adding the area column at load time? If that latter, I am concerned that this would slow down loading of datasets if not using them for visualization.

measty replied:
At the moment, the behaviour is that if you try to make a query that filters by area, and there isn't an area column, it raises an error stating that there is no area column.

Ordering by area happens by default if the area column is available. If it's not no ordering happens. At the moment there is no ability to specify what ordering happens. Probably in 2.0 we could add support for specifying what to sort on, be it area or perhaps some property.

measty commented:
> I've make a few comments / suggestions of things. Mostly just suggesting alternative naming for things.

All the comments have been addressed now, I think
